### PR TITLE
Add a source generator for a Maths language to the samples

### PR DESCRIPTION
**Pull request by lucabol**
The scenario is that a financial/gas/whatever-scientific company wants their mathematicians to write their formulas with a syntax familiar to them and without the risk of introducing bugs. But they also need to run as fast as compiled code (i.e. inline), because that is what their business relies on.

Enter source generators. They now can write:

```
AreaSquare(l)       = pow(l, 2)
AreaRectangle(w, h) = w * h
AreaCircle(r)       = pi * r * r
Quadratic(a, b, c)  = {-b + sqrt(pow(b,2) - 4 * a * c)} / (2 * a)

PointDist(x', x'', y', y'')     = sqrt(pow([x'-x''],2) + pow([y'-y''], 2))
```

I needed to build a lexer / parser for the language instead of simple regex substitution to avoid the possibility of an injection attack by a disgruntled employee. I built it in the simplest possible way for the sake of example.

The language supports:
1. Using prime char in identifiers
2. Using differents parenthesis
3. Calling functions in the Math class lowercase and functions defined before in the file (no other functions allowed)
4. Simpler syntax than C# code (i.e. no types etc...)

This is not completed yet, but done a pull request for the sake of review.

### CODE REVIEW

File: samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace ___MathLibrary___ {
+ public static partial class Formulas {
+
+        public static IEnumerable<double> ConvertToDouble(IEnumerable<int> col)
+        {
+            foreach (var s in col)
+                yield return (double)Convert.ChangeType(s, typeof(double));
```
Comment (jaredpar) [inline samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs:424]: Is there a reason that you didn't just use `yield return (double)s` here? The conversion from `int` to `double` is non-lossy here, can't see the need for `ChangeType`

File: samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    public class MathsGenerator : ISourceGenerator
+    {
+        private bool libraryIsAdded = false;
+
+        private const string libraryCode = @"
+using System.Linq;
+using System;
+using System.Collections.Generic;
+
+namespace ___MathLibrary___ {
```
Comment (jaredpar) [inline samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs:418]: Hmm, I think I get what you're trying to do here (avoid conflicting with user defined types) and yeah we don't have hygenic macros / names in the language. Same time the `__` in the name throws me off a bit. 

Is there a reason you just didn't make it a util style class, like `FormulaHelpers`, inside the `Math` namepace that you're generating?

**Review by jaredpar — APPROVED** [review]

File: samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs (hunk 1 above)
Comment (lucabol) [reply samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs]: Fixed in 8fbc804e9af50d2f6bd4b8f5b24962f236386853 . Not sure why I wrote it that way originally.

File: samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs (hunk 2 above)
Comment (lucabol) [reply samples/CSharp/SourceGenerators/SourceGeneratorSamples/MathsGenerator.cs]: Fixed in 425941d959a4c69588f79e7407841ab1e97fba0a. Great point about hygienic macros. I didn't realize the similarity. I need to remember to touch on that in the blog post.

**Review by jaredpar — APPROVED** [review]

**Comment by jmarolf** [comment]
@lucabol can you pull in the latest changes from master? `SourceGeneratorContext` no longer exists in the public API and all of our samples where updated here: https://github.com/dotnet/roslyn-sdk/pull/625

**Comment by lucabol** [comment]
I am closing this PR as it was created from an old fork. #695  represents the new one we are merging.

**Pull request closed without merging**